dvcfs: initial docs (#3932)
iterative/dvc.org: content/docs/api-reference/dvcfs.md

# DvcFileSystem

> DvcFileSystem provides a pythonic file interface (aka

Reviewer: aka -> fsspec-compatible

> DvcFileSystem provides a pythonic file interface (aka

Reviewer 2: Maybe it can be further simplified to stress the main points that it's read-only and fsspec comapitable. I think it should start with something like, "DVCFileSystem is a read-only [fsspec](https://filesystem-spec.readthedocs.io/)-compatible interface." Then you could go on to describe the typical operations and everything else.

Author: Initially I wrote like that, but I later thought that the end users does not care about fsspec or need to care about it to use it.

Reviewer 3: The `read-only` part is important

Reviewer 2: > Initially I wrote like that, but I later thought that the end users does not care about fsspec or need to care about it to use it.

Good point, but if we are not documenting the available methods, we need to point out clearly that it implements the read-only methods available in fsspec.

Reviewer 2: If you want to first focus on it being a filesytem-like interface and read-only, you could say something like: "DvcFileSystem provides read-only operations like `ls`, `du`, `glob`, `get`, `download`, etc. for a DVC repo. The API implements most read-only methods of [fsspec](https://filesystem-spec.readthedocs.io/)."

Reviewer 4: > the end users does not care about fsspec or need to care about it to use it.

So why do we even mention it?

Author: It is fsspec-based filesystem. We do need to mention it, I just don't think we should lead with that in the intro. 🙂

Reviewer 4: Sounds good. Please see https://github.com/iterative/dvc.org/pull/3932#issuecomment-1246342519 below. Thanks

> [fsspec](https://filesystem-spec.readthedocs.io/)) for a DVC repo and provides a

Reviewer: It implements read-only operations like ...

Reviewer: (too many "provides")

read-only filesystem-like operations like `ls`, `du`, `glob`, `get`, `download`,
etc.

DvcFileSystem provides a single view of all the files/directories in your
repository, be it Git-tracked or DVC-tracked, or untracked (in case of a local

> repository). DvcFileSystem is smart to reuse the files in your cache directory

Reviewer: Internally it avoids downloading DVC-tracked files unless it's needed. It's using DVC cache [make it abbr] to avoid downloading objects every time and support streaming for supported remote storages [link] for operations like ...

(if present) and can otherwise stream/fetch the files from your default remote.

> ### Basic Usage

Reviewer: do we need this subtitle?

```py

> >>> from dvc.fs.dvc import _DvcFileSystem as DvcFileSystem

Author: This is temporary. Eventually, this will look something like:
```suggestion
>>> from dvc.api import DvcFileSystem
```

Or, better:
```suggestion
>>> from dvc.fs import DvcFileSystem
```

Reviewer 2: When do we expect this change?

Author: This week. It’s not difficult, we just need to decide.

I find `dvc.fs` namespace more natural, but it’s controversial compared to `dvc.api`.
Let me know if you have any opinions/preference.

Reviewer 2: No strong preference as long as it's not `_DvcFileSystem`!

Reviewer 4: Wouldn't this be capitalized `DVCFileSystem`?

Reviewer: yep, probably should be ... for consistency with DVCLive

Author: In camel casing, the abbreviations are usually treated the same as lowercase words.

From [Wikipedia/Camel_case#Programming_and_coding](https://en.wikipedia.org/wiki/Camel_case#Programming_and_coding):
> Programming identifiers often need to contain [acronyms and initialisms](https://en.wikipedia.org/wiki/Acronyms_and_initialisms) that are already in uppercase, such as "old HTML file". By analogy with the title case rules, the natural camel case rendering would have the abbreviation all in uppercase, namely "oldHTMLFile". However, this approach is problematic when two acronyms occur together (e.g., "parse DBM XML" would become "parseDBMXML") or when the standard mandates lower camel case but the name begins with an abbreviation (e.g. "SQL server" would become "sQLServer"). For this reason, some programmers prefer to treat abbreviations as if they were lowercase words and write "oldHtmlFile", "parseDbmXml" or "sqlServer".

Also, [Google Javascript Style Guide](https://google.github.io/styleguide/jsguide.html#naming-camel-case-defined):
> lowercase everything (including acronyms), then uppercase only the first character of:
> a) … each word, to yield upper camel case, or
> b) … each word except the first, to yield lower camel case

Reviewer: Good points, @skshetry how do you feel about consistency though? (DVCLive). I bothers me a bit. Are there popular examples in Python world with all upper acronyms?

Author: MLFlow?

Author: [`socketserver.TCPServer`](https://docs.python.org/3/library/socketserver.html#socketserver.TCPServer) and [`http.HTTPServer`](https://docs.python.org/3/library/http.server.html#module-http.server) comes to mind from Python.

Author: Also urllib's HTTPResponse (Django uses `HttpResponse`). Also `json.JSONDecoder`. But `HTTP`/`JSON`/`TCP` are common acronyms.

Author: [PEP8](https://peps.python.org/pep-0008/#descriptive-naming-styles):
> Note: When using acronyms in CapWords, capitalize all the letters of the acronym. Thus HTTPServerError is better than HttpServerError.

[Pocoo Styleguide: flask/click, etc.](https://flask.palletsprojects.com/en/1.1.x/styleguide/#naming-conventions):
> Class names: CamelCase, with acronyms kept uppercase (HTTPWriter and not HttpWriter)

Reviewer 3: Afaik, all fsspec implementations (which is Python) are upper case when using acronyms: `SSHFileSystem`, `GCSFileSystem`,  `IPFSFileSystem`, `OSSFileSystem`, etc

Reviewer 3: > Are there popular examples in Python world with all upper acronyms?

scikit-learn has tons of examples using acronyms with all caps: `PLSCanonical`, `KNNImputer`, `RBFSampler`, `SGDClassifier`, `RANSACRegressor`, etc.

# opening a local repository

> >>> fs = DvcFileSystem("/path/to/local/repository")

Reviewer 2: Does this represent the current state of the repo? It looks like `_DvcFileSystem.__init__()` expects a `Repo` instance, and it doesn't seem to work for me with a `str` arg:

```python
>>> fs = DvcFileSystem(".")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/dave/miniforge3/envs/dvc/lib/python3.10/site-packages/fsspec/spec.py", line 76, in __call__
    obj = super().__call__(*args, **kwargs)
  File "/Users/dave/Code/dvc/dvc/fs/dvc.py", line 118, in __init__
    self.hash_jobs = repo.fs.hash_jobs
AttributeError: 'str' object has no attribute 'fs'
```

Author: See https://github.com/iterative/dvc.org/pull/3932#issuecomment-1242232523.

# opening a remote repository

> >>> remote_fs = DvcFileSystem("https://github.com/iterative/example-get-started.git", rev="main")
> ```

Reviewer 3: ```suggestion
>>> remote_fs = DvcFileSystem(
    "https://github.com/iterative/example-get-started.git", 
    rev="main")
```
```
The `rev` arg gets cropped in the block and it's not visible without scrolling. I think its important to show it given that it's not present in the fsspec API

Author: Thanks, I think I'll split them into two lines rather than indenting it.

> ## Examples

Author: Instead of starting with the verbose docs, I am going with "Show me the code" approach.

Reviewer 4: For now reusing the familiar `## Example: {Title}` pattern would be a first improvement (visible in the right-hand ToC).

> ## Examples

Reviewer: let's remove this subtitle also and just go use case by use case? (if we don't have that much besides examples, we don't need this structure + we can make it everything second level, so that it's easier to navigate), wdyt?

> ### Listing all DVC-tracked files

Reviewer: suggestion: make title look like command/calls - `ls`: list DVC-tracked files

Author: Like @daavoo said, it's equivalent to `ls --recursive`, so I don't want to force commands in the title. I do mention that after the snippets now. :)

```py

> >>> fs = DvcFileSystem("https://github.com/iterative/example-get-started.git", rev="main")

Reviewer: may be worth mentioning in those examples important aspects of their behavior - e.g. it doesn't download files, streams, etc. Can be brief

>>> fs.find("/", detail=False, dvc_only=True)
[
    '/data/data.xml',
    '/data/features/test.pkl',
    '/data/features/train.pkl',
    '/data/prepared/test.tsv',
    '/data/prepared/train.tsv',
    '/evaluation/importance.png',
    '/model.pkl'
]
```

### Downloading a file or a directory

```py
>>> fs = DvcFileSystem("https://github.com/iterative/example-get-started.git", rev="main")
>>> fs.get_file("data/data.xml", "data.xml")
```

This downloads "data/data.xml" file to the current working directory as
"data.xml" file. The DVC-tracked files may be downloaded from the cache if it
exists or may get streamed from the remote.

```py
>>> fs = DvcFileSystem("https://github.com/iterative/example-get-started.git", rev="main")
>>> fs.get("data", "data", recursive=True)
```

This downloads all the files in "data" directory - be it Git-tracked or
DVC-tracked into a local directory "data". Similarly, DVC might fetch files from
remote if they don't exist in the cache.

> ## API Reference

Author: I am not really seeing much value from "API reference" right now compared to Scenarios/Recipes that we can cover.

Reviewer 4: Goes back to https://github.com/iterative/dvc.org/issues/3927#issuecomment-1245044754. If we have the API ref auto-generated then I agree. Or do you suggest skipping it completely and let people/IDEs check docstrings?

Reviewer: fsspec is supposed to be a documented API that this library implements, we don't want to repeat everything here. Also, it makes sense to start small, with examples. I hope a lot of things here will be self descriptive through examples.

As DvcFileSystem is based on [fsspec](https://filesystem-spec.readthedocs.io/),

> it is compatible with most of the APIs that it offers. Please check the fsspec's

Reviewer 2: ```suggestion
it is compatible with most of the read-only methods that it offers. Please check the fsspec's
```

Author: It's not the API that is read-only, but the filesystem and it's operation in general. I have also myself struggled to explain this, but tbh I don't think we need to mention anything other than "... is a read-only filesystem".

Author: I think it's better to even skip it and raise a `EROFS`.

Reviewer 2: What is an `EROFS`?

I must be missing your point here because I don't see what's confusing or ambiguous. Is dvcfs compatible with the read-only methods of https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.spec.AbstractFileSystem?

Author: EROFS is an errno which is raised for "Read Only FileSystem". What I am trying to say is that the filesystem in general is read-only, not the methods. Take `fs.open()` for example. It works with `mode="rb"` but `mode="wb"` raises `EROFS`. So I am trying to distinguish between the method `open` vs the operations: writing/reading, etc.
